systems/plants/@RigidBodyManipulator/manipulatorDynamics.m

```diff
@@ -1,221 +1,515 @@
+      dC(:,m.NB+1:end) = dC(:,m.NB+1:end)+ diag(fc_drv);
+    else
+      [H,C] = HandC(m,q,qd,f_ext,obj.gravity);
     end
     
-    % minor adjustment to make TaylorVar work better.
-    %H = zeros(m.NB);
-    H=zeros(m.NB)*q(1);
-    
-    %Derivatives wrt q(k)
-    dH = zeros(m.NB^2,2*m.NB)*q(1);
-    for k = 1:m.NB
-      nk = m.position_num(k);
-      for i = 1:m.NB
-        n = m.position_num(i);
-        fh = IC{i} * S{i};
-        dfh = dIC{i,nk} * S{i};  %dfh/dqk
-        H(n,n) = S{i}' * fh;
-        dH(n + (n-1)*m.NB,nk) = S{i}' * dfh;
-        j = i;
-        while m.parent(j) > 0
-          if j==k,
-            dfh = Xup{j}' * dfh + dXupdq{j}' * fh;
-          else
-            dfh = Xup{j}' * dfh;
-          end
-          fh = Xup{j}' * fh;
-          
-          j = m.parent(j);
-          np = m.position_num(j);
-          
-          H(n,np) = S{j}' * fh;
-          H(np,n) = H(n,np);
-          dH(n + (np-1)*m.NB,nk) = S{j}' * dfh;
-          dH(np + (n-1)*m.NB,nk) = dH(n + (np-1)*m.NB,nk);
-        end
+    C=C + computeFrictionForce(obj,qd);
+  end
+end
+
+end
+
+function [H,C,B,dH,dC,dB] = manipulatorDynamicsNew(obj,q,v,use_mex)
+% manipulatorDynamics  Calculate coefficients of equation of motion.
+% [H,C,B,dH,dC,dB] = manipulatorDynamics(obj,q,v,use_mex) calculates the
+% coefficients of the joint-space equation of motion,
+% H(q)*vd+C(q,v,f_ext)=B(q,v)*tau, where q, v and vd are the joint
+% position, velocity and acceleration vectors, H is the joint-space inertia
+% matrix, C is the vector of gravity, external-force and velocity-product
+% terms, and tau is the joint force vector.
+%
+% Algorithm: recursive Newton-Euler for C, and Composite-Rigid-Body for H.
+
+compute_gradients = nargout > 3;
+
+options.use_mex = use_mex;
+options.compute_gradients = compute_gradients;
+options.compute_JdotV = true;
+kinsol = doKinematics(obj, q, v, options);
+
+% if (nargin<4) use_mex = true; end
+if compute_gradients
+  [f_ext, B, df_ext, dB] = computeExternalForcesAndInputMatrix(obj, q, v);
+else
+  [f_ext, B] = computeExternalForcesAndInputMatrix(obj, q, v);
+end
+
+a_grav = [zeros(3, 1); obj.gravity];
+if (use_mex && obj.mex_model_ptr~=0 && isnumeric(q) && isnumeric(v))
+  f_ext = full(f_ext);  % makes the mex implementation simpler (for now)
+  if compute_gradients
+    df_ext = full(df_ext);
+    [H, dH] = massMatrixmex(obj.mex_model_ptr);
+    nv = obj.num_velocities;
+    dH = [dH, zeros(numel(H), nv)];
+    [C, dC] = inverseDynamicsmex(obj.mex_model_ptr, f_ext, [], df_ext);
+  else
+    H = massMatrixmex(obj.mex_model_ptr);
+    C = inverseDynamicsmex(obj.mex_model_ptr, f_ext);
+  end
+  % TODO: implement in mex
+  if compute_gradients
+    [tau_friction, dtau_frictiondv] = computeFrictionForce(obj, kinsol.v);
+    nq = obj.num_positions; nv = obj.num_velocities;
+    dC(:, nq + (1:nv)) = dC(:, nq + (1:nv)) + dtau_frictiondv;
+  else
+    tau_friction = computeFrictionForce(obj, kinsol.v);
+  end
+  C = C + tau_friction;
```

Comment:
I hadn't counted on needing to implement friction in mex. This passes testManipulatorDynamics when use_new_kinsol = true but is slow. Is it OK to add this to my list of cleanup things for the next pull request?